### ISSUE TITLE
Building cockroach with Xcode 7.0/clang 7.0 fails

### ISSUE THREAD

**Issue by tildeleb** [cockroachdb/cockroach#2647]
Overview
-------------
Over the weekend I tried to build CockroachDB using Xcode 7 and it fails to build. The C++ components are the issue and they get error messages such as:

    test.cpp:1:10: fatal error: 'unordered_map' file not found

I recently upgraded to Yosemite 10.10.5 and Xcode 7.0 after being on 10.8 and Xcode 3.0 for a long time. When I say Xcode I really mean clang as I don't use the GUI much at all.

    leb@hula:~/gotest/src/github.com/cockroachdb/c-protobuf %  clang --version
    Apple LLVM version 7.0.0 (clang-700.0.72)
    Target: x86_64-apple-darwin14.5.0
    Thread model: posix
    leb@hula:~/gotest/src/github.com/cockroachdb/c-protobuf % 

C++ build failures with Xcode 7
---------------------------------------
I did some searching and I believe the fix is pretty simple. Adding `-stdlib=libc++` to *both* CFLAGS and LDFLAGS seems to work.

One thing I don't have an easy way to check out is if this flag is backwards compatible with older versions of clang and different versions of gcc on darwin.

The fixes are easy, for example, the fix for c-proto is as simple as:

    leb@hula:~/gotest/src/github.com/cockroachdb/c-protobuf % git diff
    diff --git a/cgo_flags.go b/cgo_flags.go
    index 6f2781c..06ef2b3 100644
    --- a/cgo_flags.go
    +++ b/cgo_flags.go
    @@ -3,6 +3,7 @@
     // in.
     package protobuf
 
    -// #cgo CXXFLAGS: -std=c++11
    +// #cgo CXXFLAGS: -std=c++11 -stdlib=libc++
    // #cgo CPPFLAGS: -DHAVE_CONFIG_H -Iinternal/src
    +// #cgo LDFLAGS: -stdlib=libc++
    import "C"
    leb@hula:~/gotest/src/github.com/cockroachdb/c-protobuf % ``

For RocksDB I modified:

    leb@hula:~/gotest/src/github.com/cockroachdb/c-rocksdb % git diff
    diff --git a/cgo_flags.go b/cgo_flags.go
    index a6b048b..caa0685 100644
    --- a/cgo_flags.go
    +++ b/cgo_flags.go
    @@ -10,8 +10,8 @@ package rocksdb
     // #cgo darwin CPPFLAGS: -DOS_MACOSX
     // #cgo linux CPPFLAGS: -DOS_LINUX
     // #cgo CXXFLAGS: -W -Wextra -Wall -Wsign-compare -Wshadow -Wno-unused-parameter
    -// #cgo CXXFLAGS: -std=c++11 -fno-omit-frame-pointer -momit-leaf-frame-pointer
    +// #cgo CXXFLAGS: -std=c++11 -stdlib=libc++ -fno-omit-frame-pointer -momit-leaf-frame-pointer
     // #cgo darwin CXXFLAGS: -Wshorten-64-to-32
    -// #cgo darwin LDFLAGS: -Wl,-undefined -Wl,dynamic_lookup
    +// #cgo darwin LDFLAGS: -Wl,-undefined -Wl,dynamic_lookup -stdlib=libc++
     // #cgo !darwin LDFLAGS: -Wl,-unresolved-symbols=ignore-all
     import "C"

and for cockroach/storage/engine:

    leb@hula:~/gotest/src/github.com/cockroachdb/cockroach/storage/engine % git diff
    diff --git a/storage/engine/rocksdb/rocksdb.go b/storage/engine/rocksdb/rocksdb.go
    index bfbe075..100451b 100644
    --- a/storage/engine/rocksdb/rocksdb.go
    +++ b/storage/engine/rocksdb/rocksdb.go
    @@ -26,10 +26,10 @@ import (
     )
 
     // #cgo CPPFLAGS: -I ../../../../c-protobuf/internal/src -I ../../../../c-rocksdb/internal/include
    -// #cgo CXXFLAGS: -std=c++11
    +// #cgo CXXFLAGS: -std=c++11 -stdlib=libc++
     // #cgo darwin LDFLAGS: -Wl,-undefined -Wl,dynamic_lookup
     // #cgo !darwin LDFLAGS: -Wl,-unresolved-symbols=ignore-all
    -// #cgo linux LDFLAGS: -lrt
    +// #cgo linux LDFLAGS: -lrt -stdlib=libc++
     import "C"
 
     // Logger is a logging function to be set by the importing package. Its
    leb@hula:~/gotest/src/github.com/cockroachdb/cockroach/storage/engine %

similar changes in:
cockroachdb/c-protobuf/cmd
cockroachdb/c-protobuf/libprotoc

After making these changes I get a clean build and a `cockroach` binary but I get the following `dyld` error when I try to run it:

    leb@hula:~/gotest/src/github.com/cockroachdb/cockroach % ./cockroach 
    dyld: Symbol not found: __ZNSs4_Rep11_S_max_sizeE
    Referenced from: /Volumes/std/usr/leb/gotest/src/github.com/cockroachdb/cockroach/./cockroach
    Expected in: flat namespace
    in /Volumes/std/usr/leb/gotest/src/github.com/cockroachdb/cockroach/./cockroach
    [1]    92242 trace trap  ./cockroach
    leb@hula:~/gotest/src/github.com/cockroachdb/cockroach % 

The damangled name of that symbol is `_std::string::_Rep::_S_max_size`. I've tried searching for that symbol using a recursive `find` with `nm` piped to `grep` but I haven't found which file references the symbol yet. One of the minor downsides of using `cgo` to do the builds for C++ components is the .o files aren't left around to scan so now I have modify the regular builds in the internal directories to work as well.

Anyway, I figured now would be a good time to post what I've done. I'm not up to speed on the current C++ library issues and newest C++ standards and why a switch is now needed for the C++ standard library. 

Anyone have any thoughts?

Is anyone else using Xcode 7.0?

Minor Makefile issue with glock
-----------------------------------------
BTW, I ran into a small bug in the makefile with glock. I got the following error:
    golang.org/x/net                                   7654728e3819	[OK]
    golang.org/x/text                                  505f8b49cc14	[OK]
    # cd /Volumes/std/usr/leb/gotest/src/golang.org/x/tools; hg update -r b48dc8da98ae
    abort: unknown revision 'b48dc8da98ae'!
    exit status 255
    go build -tags ''  -ldflags '-X "github.com/cockroachdb/cockroach/util.buildTag=alpha-4947g705e7d0-dirty" -X "github.com/cockroachdb/cockroach/util.buildTime=2015/09/23 

The issue was I had an older hg (not git) repository glock fails because it seems to know about hg but the supplied hash is for the git respository. This is glock issue. More importantly the Makefile doesn't stop the build when glock gets an error and it probably should?

Minor Makefile issue, `make clean` sometimes does `make build`
-----------------------------------------------------------------------------------
`make clean` sometime does a `make build`. I am not sure why this is. I think is has something to do with an interrupted build because after a build completes properly `make clean` works fine. Here is an actual log from my machine with some lines elided:

    leb@hula:~/gotest/src/github.com/cockroachdb/cockroach % make clean
    github.com/agtorre/gocolorize                      f42b554bf7f0	[OK]
    github.com/biogo/store                             e1f74b3c58be	[OK]
    github.com/cockroachdb/c-lz4                       6e71f140a365	[OK]
    ...
    golang.org/x/tools                                 b48dc8da98ae	[OK]
    gopkg.in/yaml.v1                                   9f9df34309c0	[OK]
    cmd github.com/cockroachdb/c-protobuf/cmd/protoc               	[built]
    ...
    cmd golang.org/x/tools/cmd/stringer                            	[built]
    go build -tags ''  -ldflags '-X "github.com/cockroachdb/cockroach/util.buildTag=alpha-4947-g705e7d0-dirty" -X "github.com/cockroachdb/cockroach/util.buildTime=2015/09/23 19:30:27" -X "github.com/cockroachdb/cockroach/util.buildDeps=github.com/biogo/store:e1f74b3c58befe661feed7fa4cf52436de753128 github.com/cockroachdb/c-lz4:6e71f140a365017bbe0904710007f8725fd3f809 github.com/cockroachdb/c-protobuf:6a18bfcdd5169966cd1235edb71d749c950216e8
     ...
    golang.org/x/text:505f8b49cc14d790314b7535959a10b87b9161c7 gopkg.in/yaml.v1:9f9df34309c04878acc86042b16630b0f696e1de"' -v -i -o cockroach
    github.com/cockroachdb/cockroach

I didn't track this one down. It's kind of a pain because you have to wait for the build to complete before you can get `make clean` to work.

**Comment by tamird** [comment]
Easy stuff first:
> `make clean` sometime does a `make build`

I can sort of reproduce this locally - `make clean` will nuke everything under the cockroach org, which means the next time you `bootstrap` (`make <anything>`), you're rebuilding all kinds of things. Perhaps it's time to remove `make clean`.

> The issue was I had an older hg (not git) repository glock fails because it seems to know about hg but the supplied hash is for the git respository

The is probably not glock's fault - this looks like a `go get` issue to me. Not much to be done about this. You could file an issue [upstream](https://github.com/golang/go).

> C++ build failures with Xcode 7

Those issues are pretty bizarre. FWIW, I'm also using XCode 7.0:
```shell
$ clang --version
Apple LLVM version 7.0.0 (clang-700.0.72)
Target: x86_64-apple-darwin14.5.0
Thread model: posix
```

That you're getting a missing symbol on [`std::string::max_size`](http://www.cplusplus.com/reference/string/string/max_size/) seems to mean that your system's libc++ is not in the right place.

I haven't rebuilt any of our c++ dependencies since upgrading to XCode 7.0, but I've read that XCode 7.0 doesn't ship with the OS X SDK. Can you try installing it (`xcode-select --install`) and seeing if that fixes your issue (without the patches you listed)?

**Comment by tildeleb** [comment]
@tamird thanks for the help and the quick response.

FWIW, I can't do `xcode-select --install` as I get this error:

    leb@hula:~/gotest/src/github.com % xcode-select --install
    xcode-select: error: command line tools are already installed, use "Software Update" to install updates
    leb@hula:~/gotest/src/github.com % 

After I upgraded to Yosemite I installed Xcode 7.0 via the App Store and then I went to the Apple Developer web site and downloaded and installed the command line tools for Xcode 7 and OS X 10.10 e.g. `Command_Line_Tools_OS_X_10.10_for_Xcode_7.dmg`

So I don't think there is anything strange about how I installed Xcode. 

You said you are running Xcode 7.0. Are you running 10.10.5?

As far as the system's libc++ not being in the right place. I have the following test program:

    leb@hula:~/Desktop/work % cat test.cpp
    #include <iostream>
    #include <unordered_map>
    #include <string>

    int main() {
        size_t siz, len, cap, max; 
        std::string str = "Hello World";
        siz = str.size(); len = str.length(); cap = str.capacity(); max = str.max_size();
        std::cout << "size: " << siz << std::endl; std::cout << "len: " << len << std::endl;
        std::cout << "cap: " << cap << std::endl; std::cout << "max: " << max << std::endl;
        
        std::unordered_map <int, int> map;
        std::cout << "map size: " << map.size () << std::endl;
        return 0;   
    }
    leb@hula:~/Desktop/work % clang++ -std=c++11 -stdlib=libc++ test.cpp
    leb@hula:~/Desktop/work % a.out                                     
    size: 11
    len: 11
    cap: 22
    max: 18446744073709551599
    map size: 0
    leb@hula:~/Desktop/work % 

So at some level things are working and a C++ program can access libc++ and can access str.max_size?

The name of the symbol is not ` std::string::max_size` but`_ZNSs4_Rep11_S_max_sizeE`
which I think becomes `_std::string::_Rep::_S_max_size`.

I am not sure the demangler worked correctly but what is "Rep::"?

**Comment by tamird** [comment]
Does your test program compile without the `-stdlib=libc++` flag? It does on my machine. In fact, it compiles on my machines with just `clang++ test.cpp`.

For some reason, `clang` is defaulting to `libstdc++` on your machine, even though the default has been `libc++` since OS X 10.9. What does `xcrun --show-sdk-version` print on your machine?

**Comment by tildeleb** [comment]
@tamird Again thanks for the help and the quick responses. 

FWIW, I skipped 10.9 and upgraded from 10.8 to 10.10. I haven't done any C++ for a long time and I wasn't aware that Apple now has their own version of the C++ standard library.

I wonder if there is any difference between `xcode-select --install` and the command line tools dmg image for Xcode 7/10.10?

To answer your question the test program doesn't compile without the -stdlib=libc++ flag.

    leb@hula:~/Desktop/work % clang++ test.cpp
    test.cpp:2:10: fatal error: 'unordered_map' file not found
    #include <unordered_map>
             ^
    1 error generated.
    leb@hula:~/Desktop/work % 

What I don't really understand is shouldn't the `#include <unordered_map>` work no matter which library is selected? To me this seems like a library search path or bad symbolic link to a library kind of issue?

    leb@hula:~ % xcrun --show-sdk-version
    10.11
    leb@hula:~ % 

Nice idea, if the SDK was old that might explain it but it's 10.11, right?

What do you get for this?

    leb@hula:/usr/lib % ls -l /usr/lib/{libc\+\+*,libstd*}
    -rwxr-xr-x  1 root  wheel  1414480 Sep  9  2014 /usr/lib/libc++.1.dylib*
    lrwxr-xr-x  1 root  wheel       14 Sep  5 21:13 /usr/lib/libc++.dylib@ -> libc++.1.dylib
    -rwxr-xr-x  1 root  wheel   441136 Sep  9  2014 /usr/lib/libc++abi.dylib*
    -rwxr-xr-x  1 root  wheel  1459408 Sep  9  2014 /usr/lib/libstdc++.6.0.9.dylib*
    lrwxr-xr-x  1 root  wheel       21 Sep  5 21:13 /usr/lib/libstdc++.6.dylib@ -> libstdc++.6.0.9.dylib
    lrwxr-xr-x  1 root  wheel       17 Sep  5 21:13 /usr/lib/libstdc++.dylib@ -> libstdc++.6.dylib
    leb@hula:/usr/lib % 

It says here:
http://libcxx.llvm.org

    Using libc++ in your programs

    FreeBSD and Mac OS X
    To use your system-installed libc++ with clang you can:

    clang++ -stdlib=libc++ test.cpp
    clang++ -std=c++11 -stdlib=libc++ test.cpp
    To use your tip-of-trunk libc++ on Mac OS with clang you can:

    export DYLD_LIBRARY_PATH=path/to/build/lib
    clang++ -std=c++11 -stdlib=libc++ -nostdinc++ -I<path-to-libcxx>/include -L<path-to-libcxx>/lib test.cpp

This kind of contradicts what you are saying about libc++ being the default, although I believe you. Maybe the doc is stale. But wouldn't that kind of change be tied to a version of Xcode/clang and not to a version of OS-X or the SDK selection?

So the question is why is my install of Xcode 7 different? My system is kind of ancient as it's been upgraded time after time so there's always a chance something is messed up. I have some plan9 tools in my search path and so on. I'll try to snoop around and figure it out. I have a friend at Apple who would know or could find out for me but I'd rather not bug him unless I have to.

My son has a stock system running 10.9. I think I'll try it there tomorrow while he's at school and get another data point. After that maybe upgrade it to 10.10 and Xcode 7 as well and get yet another data point.

**Is anyone else using Xcode/clang 7?**

**Comment by tamird** [comment]
> What I don't really understand is shouldn't the #include <unordered_map> work no matter which library is selected?

`unordered_map` is a c++11 thing; libc++ is the c++11 standard library.

> Nice idea, if the SDK was old that might explain it but it's 10.11, right?

Yeah, that's the same as mine.

> What do you get for this?
```shell
$ ls -l /usr/lib/{libc\+\+*,libstd*}
-rwxr-xr-x  1 root  wheel  1414480 Sep  9  2014 /usr/lib/libc++.1.dylib
lrwxr-xr-x  1 root  wheel       14 Apr 14 16:06 /usr/lib/libc++.dylib -> libc++.1.dylib
-rwxr-xr-x  1 root  wheel   441136 Sep  9  2014 /usr/lib/libc++abi.dylib
-rwxr-xr-x  1 root  wheel  1459408 Sep  9  2014 /usr/lib/libstdc++.6.0.9.dylib
lrwxr-xr-x  1 root  wheel       21 Apr 14 16:06 /usr/lib/libstdc++.6.dylib -> libstdc++.6.0.9.dylib
lrwxr-xr-x  1 root  wheel       17 Apr 14 16:06 /usr/lib/libstdc++.dylib -> libstdc++.6.dylib
```

Why do yours have trailing `*`s and `@`s?

**Comment by tildeleb** [comment]
Thanks and kudos to @tamird for getting me thinking along the lines of SDK versions tied to the version of OS X. I woke up this morning and had an epiphany. The problem was this:

    leb@hula:~ % printenv | grep MACOSX_DEPLOYMENT_TARGET
    MACOSX_DEPLOYMENT_TARGET=10.8
    leb@hula:~ % 

So the setting of that environemnt variable was changing the default behavior of clang. I set this long ago to get something else to work properly without having to add the flag (see below) to some pre-exsiting build system and I remember thinking when I set it that it's going to screw me over at some day and it did. Here is the clang doc for it for the environment variable.

    MACOSX_DEPLOYMENT_TARGET
    If -mmacosx-version-min is unspecified, the default deployment target is read from this environment variable. This option only affects Darwin targets.

As the doc says there is also a `-mmacosx-version-min` flag that does the same thing or something similar:

    -mmacosx-version-min=<version>
    When building for Mac OS X, specify the minimum version supported by your application.

FWIW I thought this environment variable controlled which of several SDKs clang selected. I had no idea it would change things like the default values of clang's flags. BTW I ran `clang++ -print-search-dirs` and it prints out the same thing with and without the `MACOSX_DEPLOYMENT_TARGET`environment variable set. What it prints out doesn't even include `/usr/lib` which is clearly searched.

    leb@hula:~ % clang++ -print-search-dirs
    programs: =/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin:/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
    libraries: =/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/7.0.0
    leb@hula:~ % 

So the test program compiles now without any flags to clang, I get a clean build of cockroach, it runs, and gives me a usage message.

*I'll leave it to someone else to decide if there should be some doc changes to warn others of this behavior. Maybe this is too obscure to warrant a doc change.*
 
**Why do yours have trailing *s and @s?**
I have the following alias set in my shell.

    alias ls='/bin/ls -F'
